### DISCUSSION TITLE
如何在用lua运行脚本时传递参数？

### DISCUSSION THREAD

**Discussion by x1244** [xmake-io/xmake#5824] [Q&A]
我尝试着运行`xmake lua sth.lua abc bbc`来向脚本传递参数，但却没有得到想要的参数。
比如我有一个本地脚本`hello.lua`
```
function main(argv, opt)
    for _, value in ipairs(argv) do
        print("argv:", value)
    end
    opt = opt or {}
    print("opt:", opt)
end
```
当我运行xmake lua，获得argv值为abc，opt值为bbc
```
p> xmake lua hello.lua abc bbc
argv: abc
opt: bbc
```
当我想要给opt一个配置项时，根据`xmake lua --help`指示的用`-d`，却没有得到想要的结果，此时opt是空的
```
> xmake lua -d# hello.lua abc #{bbc=true}
argv: abc
opt: { }
```
那么我应该如何在调试类似`find_qt.lua`这样的脚本时，从命令行输入配置项opt呢？

**Comment by waruqi** [comment]
```lua
function main(...)
    local argv = {...}
end
```

**Comment by x1244** [comment]
那像`find_qt.lua`这种原生脚本，要通过`xmake lua`运行调试的时候，如何传递参数呢？
![image](https://github.com/user-attachments/assets/587257c1-908e-4450-ada4-35e6d4604349)
直接运行，不能获得opt
```
> xmake lua -d# find_qt.lua c:/ #{version="5.13"}
source: c:/
opt: { }
```
还请给指点一下

**Comment by x1244** [comment]
那像`find_qt.lua`这种原生脚本，要通过`xmake lua`运行调试的时候，如何传递参数呢？
![image](https://github.com/user-attachments/assets/587257c1-908e-4450-ada4-35e6d4604349)
直接运行，不能获得opt
```
> xmake lua -d# find_qt.lua c:/ #{version="5.13"}
source: c:/
opt: { }
```
还请给指点一下

**Reply by waruqi** [reply]
写个 test.lua 里面走 import("detect.sdks.find_qt") 调，或者直接编译 qt 程序时候，直接调 xmake 自身源码

**Comment by x1244** [comment]
这样的话，我是基本上明白了。
那能否再请解释一下`xmake lua --help`中 `-d --deserialize`的用法
```
Command options (lua):
    -d DESERIALIZE, --deserialize=DESERIALIZE
e.g.
   - xmake lua -d@ lib.detect.find_tool tar @{version=true}
```
如果我是建立的`test.lua`测试脚本，那如何获取到{version=true}参数配置呢？
注意，这里在windows下，不能用`-d@`，可以用`-d#`

**Reply by waruqi** [reply]
```console
$ xmake l -d# print arg1 "#{version=true}" arg3
arg1 {
  version = true
}
 arg3
```

**Comment by x1244** [comment]
这下更清楚了，原来传递表参数的时候要放入到引号里。
但是`xmake lua --help`里给出的示例并没有用引号括起来，导致我没有能够成功。
在将参数放入到引号里，像`hello.lua`或`find_qt.lua`这样的脚本就不用再创建一个辅助的`test.lua`脚本来传参数，可以直接在命令行传递的。
```
> xmake lua -d# hello.lua arg1 "#{version='0.1.0'}"
argv: arg1
opt: {
  version = "0.1.0"
}
```
能否请将`xmake lua --help`帮助说明更新一下，避免引起误解误用呢？
感谢支持！
